File: src/andytipster-client/src/app/core/services/audit.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface AuditLogRequest {
  page?: number;
  pageSize?: number;
  search?: string;
  actionTypeFilter?: string;
  actorFilter?: string;
  targetEntityFilter?: string;
  dateFrom?: string;
  dateTo?: string;
  sortBy?: string;
  sortDirection?: string;
}

export interface AuditLogEntry {
  id: number;
  actorUserId: string;
  actorName: string;
  actionType: string;
  targetEntity: string;
  targetEntityId: string | null;
  beforeJson: string | null;
  afterJson: string | null;
  timestamp: string;
  ipAddress: string | null;
}

export interface AuditLogResponse {
  entries: AuditLogEntry[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

@Injectable({ providedIn: 'root' })
export class AuditService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/api/audit`;

  getAuditLogs(request: AuditLogRequest): Observable<AuditLogResponse> {
    let params = new HttpParams();
    if (request.page) params = params.set('page', request.page);
    if (request.pageSize) params = params.set('pageSize', request.pageSize);
    if (request.search) params = params.set('search', request.search);
    if (request.actionTypeFilter) params = params.set('actionTypeFilter', request.actionTypeFilter);
    if (request.actorFilter) params = params.set('actorFilter', request.actorFilter);
    if (request.targetEntityFilter) params = params.set('targetEntityFilter', request.targetEntityFilter);
    if (request.dateFrom) params = params.set('dateFrom', request.dateFrom);
    if (request.dateTo) params = params.set('dateTo', request.dateTo);
    if (request.sortBy) params = params.set('sortBy', request.sortBy);
    if (request.sortDirection) params = params.set('sortDirection', request.sortDirection);

    return this.http.get<AuditLogResponse>(this.apiUrl, { params });
  }
}

File: src/AndyTipster.Infrastructure/Services/AuditService.cs
using AndyTipster.Application.Audit.Services;
using AndyTipster.Domain.Entities;
using AndyTipster.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace AndyTipster.Infrastructure.Services;

/// <summary>
/// Append-only audit logging service. Logs all admin actions with actor, target, action type,
/// timestamp, and before/after values. No UPDATE or DELETE operations are exposed.
/// Configured with 2-year retention policy (handled at database level).
/// </summary>
public class AuditService : IAuditService
{
    private readonly AndyTipsterDbContext _context;

    public AuditService(AndyTipsterDbContext context)
    {
        _context = context;
    }

    /// <inheritdoc />
    public async Task LogActionAsync(
        Guid actorUserId,
        string actionType,
        string targetEntity,
        string? targetEntityId = null,
        string? beforeJson = null,
        string? afterJson = null,
        string? ipAddress = null,
        CancellationToken cancellationToken = default)
    {
        var auditLog = new AuditLog
        {
            ActorUserId = actorUserId,
            ActionType = actionType,
            TargetEntity = targetEntity,
            TargetEntityId = targetEntityId,
            BeforeJson = beforeJson,
            AfterJson = afterJson,
            Timestamp = DateTime.UtcNow,
            IpAddress = ipAddress
        };

        _context.AuditLogs.Add(auditLog);
        await _context.SaveChangesAsync(cancellationToken);
    }

    /// <inheritdoc />
    public async Task<AuditLogResponse> GetAuditLogsAsync(AuditLogRequest request, CancellationToken cancellationToken = default)
    {
        var query = _context.AuditLogs
            .Include(a => a.ActorUser)
            .AsNoTracking()
            .AsQueryable();

        // Apply search filter (searches action type, target entity, and actor name)
        if (!string.IsNullOrWhiteSpace(request.Search))
        {
            var search = request.Search.ToLower();
            query = query.Where(a =>
                a.ActionType.ToLower().Contains(search) ||
                a.TargetEntity.ToLower().Contains(search) ||
                a.ActorUser.DisplayName.ToLower().Contains(search) ||
                (a.TargetEntityId != null && a.TargetEntityId.ToLower().Contains(search)));
        }

        // Apply action type filter
        if (!string.IsNullOrWhiteSpace(request.ActionTypeFilter))
        {
            query = query.Where(a => a.ActionType == request.ActionTypeFilter);
        }

        // Apply actor filter
        if (!string.IsNullOrWhiteSpace(request.ActorFilter))
        {
            if (Guid.TryParse(request.ActorFilter, out var actorId))
            {
                query = query.Where(a => a.ActorUserId == actorId);
            }
        }

        // Apply target entity filter
        if (!string.IsNullOrWhiteSpace(request.TargetEntityFilter))
        {
            query = query.Where(a => a.TargetEntity == request.TargetEntityFilter);
        }

        // Apply date range filter
        if (request.DateFrom.HasValue)
        {
            query = query.Where(a => a.Timestamp >= request.DateFrom.Value);
        }

        if (request.DateTo.HasValue)
        {
            query = query.Where(a => a.Timestamp <= request.DateTo.Value);
        }

        // Get total count for pagination
        var totalCount = await query.CountAsync(cancellationToken);

        // Apply sorting
        query = request.SortBy?.ToLower() switch
        {
            "actiontype" => request.SortDirection?.ToLower() == "asc"
                ? query.OrderBy(a => a.ActionType)
                : query.OrderByDescending(a => a.ActionType),
            "targetentity" => request.SortDirection?.ToLower() == "asc"
                ? query.OrderBy(a => a.TargetEntity)
                : query.OrderByDescending(a => a.TargetEntity),
            _ => request.SortDirection?.ToLower() == "asc"
                ? query.OrderBy(a => a.Timestamp)
                : query.OrderByDescending(a => a.Timestamp),
        };

        // Apply pagination
        var pageSize = Math.Clamp(request.PageSize, 1, 100);
        var page = Math.Max(request.Page, 1);
        var entries = await query
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(a => new AuditLogEntryDto
            {
                Id = a.Id,
                ActorUserId = a.ActorUserId,
                ActorName = a.ActorUser.DisplayName,
                ActionType = a.ActionType,
                TargetEntity = a.TargetEntity,
                TargetEntityId = a.TargetEntityId,
                BeforeJson = a.BeforeJson,
                AfterJson = a.AfterJson,
                Timestamp = a.Timestamp,
                IpAddress = a.IpAddress
            })
            .ToListAsync(cancellationToken);

        return new AuditLogResponse
        {
            Entries = entries,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize,
            TotalPages = (int)Math.Ceiling(totalCount / (double)pageSize)
        };
    }
}
